Heatexchanger Cleanup and Dynamic NTU Condenser Init

eslickj opened the pull request:
**NOTE:** When this PR is approved, you *must* do the following so it can be merged:
* Add the "approved" tag to the pull-request
* Trigger tests by pushing a commit or manually re-running in CircleCI

See [this wiki page](https://github.com/IDAES/idaes-dev/wiki/Github-Pull-Request-Workflow-with-Approval-Tag) for details.

----

## Summary/Motivation:
This fixes an issue with dynamic initialization of the NTU Condenser power model. Does some minor rearrangement of the heat exchanger model to improve clarity regarding side names and adds tests for both the condenser and generic heat exchanger.

### Legal Acknowledgement

By contributing to this software project, I agree to the following terms and conditions for my contribution:

1. I agree my contributions are submitted under the license terms described in the LICENSE.txt file at the top level of this directory.
2. I represent I am authorized to make the contributions and grant the license. If my employer has rights to intellectual property that includes these contributions, I represent that I have received permission to make contributions and grant the required license on behalf of that employer.

----

## Reviewer Checklist
* Reviewer 1
- [ ] Tests: do they pass, do they cover functionality, are they understandable?
- [ ] Documentation: is the code commented, are there user-level docs where needed?
* Reviewer 2
- [ ] Tests: do they pass, do they cover functionality, are they understandable?
- [ ] Documentation: is the code commented, are there user-level docs where needed?

jghouse88 commented on idaes/generic_models/unit_models/heat_exchanger.py:
```diff
@@ -162,19 +162,20 @@ def delta_temperature_underwood_callback(b):
     """
+    dT1 = b.delta_temperature_in
+    dT2 = b.delta_temperature_out
+    t_units = pyunits.get_units(dT1)
```
Suggest explicit naming like `temp_units` or `temperature_units`?

eslickj replied:
Sure.  Sounds good.  Will fix.

jghouse88 commented on idaes/generic_models/unit_models/heat_exchanger.py:
```diff
@@ -277,66 +309,67 @@ def build(self):
-        q = self.side_2.heat
+
+        self.heat_duty = Reference(cold_side.heat)
+        q = self.heat_duty
```
Can we avoid the alias for `self.heat_duty`? Might be easier to read the code instead of just `q`.

eslickj replied:
Agree, will fix.

jghouse88 commented on idaes/power_generation/unit_models/helm/condenser_ntu.py:
```diff
@@ -221,32 +233,33 @@ def build(self):
         @self.Constraint(time, doc="Heat balance equation")
         def unit_heat_balance(b, t):
-            return 0 == self.side_1.heat[t] + self.side_2.heat[t]
+            return 0 == (self.side_1.heat[t] +
```
Should this be using `hot_side` and `cold_side` instead of `side_1` and `side_2`?

eslickj replied:
Yes, although it doesn't matter.  Will fix.

jghouse88 commented on idaes/power_generation/unit_models/helm/condenser_ntu.py:
```diff
@@ -354,8 +366,8 @@ def initialize(
-        blk.side_2.release_state(flags2, outlvl + 1)
+        blk.side_1.release_state(flags1, outlvl)
+        blk.side_2.release_state(flags2, outlvl)
         from_json(blk, sd=istate, wts=sp)
```
Just a general comment that `hot_side`/`cold_side` and `side_1`/`side_2` are used interchangeably. Not sure if this is intended.

jghouse88 reviewed (changes requested):
Minor comments/questions. Looks good.

eslickj commented:
Okay.  I addressed comments.  I also added a check to make sure someone doesn't name the hot side cold_side or side_2 and doesn't name the cold side hot_side or side_1.

jghouse88 reviewed (dismissed):
LGTM. Probably better to wait for PR #54  to be merged before merging this..

andrewlee94 commented on idaes/generic_models/unit_models/heat_exchanger.py:
```diff
@@ -162,19 +163,20 @@ def delta_temperature_underwood_callback(b):
+
     # external function that ruturns the real root, for the cuberoot of negitive
     # numbers, so it will return without error for positive and negitive dT.
     b.cbrt = ExternalFunction(
```
The output of this should probably have units too; i.e. `units=pyunits.K`. This would mean you don't need to add units on line 179.

eslickj replied:
I think this is ok.  I just left the cube root function return value unitless.  I figured this would make things simpler than having units of cube root of K the get cubed back to K later.  That process made me a little uncomfortable. Like what if I ended up with units of K^0.9999999999999 or something?

eslickj replied:
@jsiirola or @carldlaird, if we have units like (K^(1/3))^3 are we sure that the resulting unit is exactly K and not just close to K?

andrewlee94 commented on idaes/generic_models/unit_models/heat_exchanger.py:
```diff
@@ -217,6 +219,11 @@ def _process_config(self):
             setattr(config, config.cold_side_name, config.cold_side_config)
 
+        if config.cold_side_name in ["hot_side", "side_1"]:
+            raise ConfigurationError("Cold side name cannot be in {'hot_side', 'side_1'}.")
```
Minor point: `{'hot_side', 'side_1'}` should be a list.

andrewlee94 commented on idaes/generic_models/unit_models/heat_exchanger.py:
```diff
@@ -217,6 +219,11 @@ def _process_config(self):
+        if config.cold_side_name in ["hot_side", "side_1"]:
+            raise ConfigurationError("Cold side name cannot be in {'hot_side', 'side_1'}.")
+        if config.hot_side_name in ["cold_side", "side_2"]:
+            raise ConfigurationError("Hot side name cannot be in {'cold_side', 'side_2'}.")
```
As above.

andrewlee94 commented on idaes/generic_models/unit_models/heat_exchanger.py:
```diff
@@ -233,16 +240,47 @@ def build(self):
+            dynamic=config.dynamic,
+            has_holdup=config.has_holdup,
+        )
+        # Add refernces to the hot side and cold side, so that we have solid
```
Type: references

andrewlee94 commented on idaes/generic_models/unit_models/heat_exchanger.py:
```diff
@@ -233,16 +240,47 @@ def build(self):
+            has_holdup=config.has_holdup,
+        )
+        # Add refernces to the hot side and cold side, so that we have solid
+        # names to refere to internally.  side_1 and side_2 also maintain
```
Typo: refer

andrewlee94 commented on idaes/generic_models/unit_models/heat_exchanger.py:
```diff
@@ -233,16 +240,47 @@ def build(self):
         u_units = s1_metadata.get_derived_units("heat_transfer_coefficient")
         a_units = s1_metadata.get_derived_units("area")
-        t_units = s1_metadata.get_derived_units("temperature")
+        temp_units = s1_metadata.get_derived_units("temperature")
```
I don't want to add more work, but it seems odd that this is `temp_units` whilst all the others are shorter now.

eslickj replied:
This was changed to address @jghouse88's comment.  I think it's just longer due to potential confusion between time and temperature.

andrewlee94 commented on idaes/power_generation/unit_models/helm/condenser_ntu.py:
```diff
@@ -109,6 +110,11 @@ def _process_config(self):
             setattr(config, config.cold_side_name, config.cold_side_config)
 
+        if config.cold_side_name in ["hot_side", "side_1"]:
+            raise ConfigurationError("Cold side name cannot be in {'hot_side', 'side_1'}.")
```
As before: `{'hot_side', 'side_1'}` should be a list.

andrewlee94 commented on idaes/power_generation/unit_models/helm/condenser_ntu.py:
```diff
@@ -109,6 +110,11 @@ def _process_config(self):
+        if config.cold_side_name in ["hot_side", "side_1"]:
+            raise ConfigurationError("Cold side name cannot be in {'hot_side', 'side_1'}.")
+        if config.hot_side_name in ["cold_side", "side_2"]:
+            raise ConfigurationError("Hot side name cannot be in {'cold_side', 'side_2'}.")
```
As above.

andrewlee94 commented on idaes/power_generation/unit_models/helm/tests/test_condenser_ntu.py:
```diff
@@ -0,0 +1,103 @@
+
+@pytest.mark.unit
+@pytest.mark.skipif(solver is None, reason="Solver not available")
+def test_condenser_steady():
```
I would like to see the tests follow the standard of the core library more, but I won't insist on that.

eslickj replied:
I'm a bit short on time right now, but hopefully very soon I'm going to update the tests and documentation for all the steam-related power models.

andrewlee94 reviewed (changes requested):
A couple of minor comments, which are mostly formatting and typos.

eslickj commented:
@andrewlee94 and @jghouse88, I addressed @andrewlee94's comment and it's ready for another look.

jghouse88 reviewed (approved):
LGTM. I am seeing that it is out of date with the main branch.

andrewlee94 reviewed (approved):
Looks good. I would like to resolve the concern about units in the cube root function, but that is not critical.
